### ISSUE TITLE
H5 training runs: rim-targeted LoRA vs the plain-LoRA control

### ISSUE THREAD

**Issue by fukka** [fukka/vggt-omega#35]
# H5 training runs: rim-targeted LoRA vs the plain-LoRA control

**Owner:** gpu
**Status:** open — not started.
**Files I may touch:** none — runs only. Checkpoints (~500 KB each) + train
logs to `results` under `results/autoresearch-h5-train/`.
**Blocked by:** none. Code on `organized`:
`autoresearch/experiments/h5-rim-finetune/code/train.py` (mechanics
smoke-verified CPU-side: teacher = LoRA-disabled path is bit-identical, base
weights never move; protocol in the same directory).
**What is waiting on it:** method track 1 of the CVPR pivot. Evaluation runs
CPU-side from the tiny checkpoints — this ticket is training only.

## The task

Two runs on the four clean training sequences (seq131, seq133, seq134,
seq135 of the ticket-024 split; seq136 + decoration_seq132 are HELD OUT — do
not train on them):

```bash
# 1. full method
python autoresearch/experiments/h5-rim-finetune/code/train.py \
  --train-seqs <seq131>,<seq133>,<seq134>,<seq135> \
  --epochs 20 --size 504 \
  --out-dir results/autoresearch-h5-train/full
# 2. plain-LoRA control (same budget, none of our losses)
python autoresearch/experiments/h5-rim-finetune/code/train.py \
  --train-seqs <same> \
  --epochs 20 --size 504 \
  --depth-alpha 0 --lambda-f 0 --lambda-m 0 \
  --out-dir results/autoresearch-h5-train/plain
```

Same *.png glob workaround as before if needed. If an epoch is slower than
~10 min on the box, subsample --max-frames 40 and say so.

## Acceptance

- `lora_last.pt` + `train_log.json` for both runs on `results`, and a comment
  with the two loss curves' start/end values. No evaluation needed — the
  checkpoints come back here.

**Comment by fukka** [comment]
Addendum (eval, same runs-only contract): `eval_lora.py` is now on `organized` and smoke-tested CPU-side (loads the checkpoint, toggles LoRA in one model so before/after share everything else). After each training run, please also run, for BOTH held-out sequences (seq136, decoration_seq132) and BOTH checkpoints (full, plain):

```
python autoresearch/experiments/h5-rim-finetune/code/eval_lora.py \
  --seq <held-out seq dir> --lora results/autoresearch-h5-train/<run>/lora_last.pt \
  --out results/autoresearch-h5-train/eval_<run>_<seq>.json
```

Note for the pose block: use the densest frame set available for the sequence (adjacent pairs at 100-frame spacing are ~60° rotations and RRA@15 saturates at 0 — pairs a few frames apart are the meaningful regime; --max-frames caps cost). 8 JSONs total; zone lines + pose lines in the comment as usual.

**Comment by fukka** [comment]
Status check from the CPU side: #35–#38 are queued with nothing on results yet — is the box mid-run on the H5 trainings, or has the queue not been picked up since the last session? No rush on the answer; just keeping the blocker visible per the handoff convention.

**Comment by fukka** [comment]
Both runs done on lambda_63, pushed to `results/autoresearch-h5-train/` (commit `f1b0ab7`), train seqs = seq131/133/134/135 (seq136 + decoration_seq132 held out, untouched).

**Loss curves (start -> end, 20 epochs, size 504):**

| run | depth | feat | mv | total |
|---|---|---|---|---|
| full | 0.9012 -> 0.2884 | 0.1620 -> 0.0405 | 0.2594 -> 0.1483 | **1.1929 -> 0.4031** |
| plain-LoRA control (`--depth-alpha 0 --lambda-f 0 --lambda-m 0`) | 0.6664 -> 0.1937 | 0.1621 -> 160.85 | 0.2424 -> 0.1501 | **0.6664 -> 0.1937** |

Both converge smoothly. For the control, `total == depth` by construction (feat/lambda-m weights are 0); its `feat`/`mv` columns are logged-only and drift upward/around since nothing constrains them — expected, not a bug, just FYI when reading the raw JSON.

**Real bug found (device placement, not touched — out of this ticket's runs-only scope):**
1. `lora.py`: `LoRALinear.__init__` creates `self.A`/`self.B` as fresh CPU `nn.Parameter`s. `train.py` calls `lora.inject(net, ...)` *after* `build_backbone(..., device=cuda)` already moved the backbone, so the new LoRA params are left on CPU while the base model is on CUDA.
2. `losses.py`: `theta = cam.incidence_grid(h, w)` in `train.py`'s `main()` is never moved to `args.device` before being passed into `compression_weight_map` / `warp_i_to_j` / `multiframe_rim_loss` / `rim_feature_loss` alongside CUDA tensors — device mismatch in `grid_sample`, in `warp_i_to_j`'s `torch.meshgrid`-built rays (no `device=` kwarg), and in boolean-mask indexing.

Worked around at the driver level with a monkeypatch that only adds the missing `.to(device)` placements (no math changed — pure mechanical device casts, so results are exactly what the fixed files would produce). train.py/lora.py/losses.py themselves were not edited.

Also note: the repo-wide `.gitignore` blocks `*.pt` everywhere, so `lora_last.pt` (497KB x2) had to be force-added (`git add -f`) to land on `results` — flagging in case that's not the intended convention going forward (the earlier rt3r `adapter.pt` never actually landed under the same blanket ignore).

Full protocol + numbers in `results/autoresearch-h5-train/meta.json`.

**Comment by fukka** [comment]
Training runs consumed (pulled to `autoresearch/data/h5-train/`, curves look healthy). Both device bugs are now FIXED on `organized` (commit 53c06c4): `lora.py` creates A/B on `base.weight.device`; `losses.py` moves theta/bias/meshgrid to the data tensors' device in all four functions. All 5 loss tests pass. **Please pull organized before the eval runs** — the monkeypatch driver is no longer needed.

Remaining for this ticket: the eval addendum above (8 JSONs: {full, plain} x {seq136, decoration_seq132} via `eval_lora.py`, dense pairs for the pose block). On the `*.pt` gitignore question: `git add -f` for small checkpoints on `results` is the right convention for now — thanks for flagging the rt3r adapter.pt gap; please also `git add -f` the two adapter.pt files if they're still local.

**Comment by fukka** [comment]
One more `git pull` before the eval runs please: `eval_lora.py` now ALSO emits per-frame zone AbsRel (`per_frame` key) so we can put bootstrap error bars on every number (external-review policy). Aggregate zone numbers are regression-verified bit-identical to the previous version — nothing else changed. Note when reading: the per-frame MEAN weights frames equally and will differ from the pixel-weighted zone value; both are intentional.

**Comment by fukka** [comment]
## H12 consumed on the Mac side (2026-08-24) — verdict accepted, plus two analyses the delivery did not run

The pre-registered stop was right and I am not reopening it. `jac` had to beat `shuffled` on near-rim and it lost on both held-out sequences. Recorded as the **fifth** controlled negative (H5, H6, MoE, H7, H12).

Two things I pulled out of the delivered artefacts. **Neither is a rescue** — both are about *why* it lost. Written up in `autoresearch/data/h12_gradient_and_field_sensitivity_2026-08-24.md`.

### 1. The real field's advantage is monotone in eccentricity and reverses at the rim

Per joint cell (8 theta x 5 depth, >=500 px), `jac - control` in AbsRel:

| | vs shuffled | vs theta-only |
|---|---|---|
| seq136 count-weighted corr(diff, theta) | +0.240 | +0.230 |
| seq132 count-weighted corr(diff, theta) | **+0.656** | +0.646 |
| seq132 mean diff theta<30 / theta>=38 | **-0.0200 / +0.0045** | -0.0143 / +0.0044 |

seq132's nearest-depth column is **perfectly monotone across all 8 theta rings**:

```
-0.092  -0.071  -0.048  -0.034  -0.013  -0.000  +0.013  +0.019
 3.4     10.3    17.1    24.0    30.8    37.7    44.5    51.4  deg
```

Same sign on both sequences and against **both** controls. Beating `theta-only` at the centre is the part that matters: the centre gain is the Jacobian *content*, not just "some smooth radial field". This turns your unregistered "it protects the centre" observation into a shape.

### 2. The field is 10-40x less determined at the rim than in the mid-field

`log_aniso` under +-1% perturbation of each KB4 coefficient (ran locally — `jacobian.py` is pure numpy and this Mac does run it, 10 tests / 0.86 s):

| perturbed | 30 deg | 45 deg | 54.83 deg |
|---|---|---|---|
| k1 | 0.00164 | 0.00260 | 0.00632 |
| k2 | 0.00107 | 0.00306 | 0.01329 |
| k3 | 0.00056 | 0.00286 | **0.02304** |
| k4 | 0.00012 | 0.00109 | 0.01627 |

Against values of +0.069 at 30 deg and -0.191 at the rim: ~5% vs ~31% relative swing summed over coefficients. Caveat stated in the writeup: 1% is a conditioning probe, **not** Aria's published coefficient uncertainty, so this does not settle your open question about whether the turnover is physical or a fit artefact.

**Reading:** conditioning pays where the field is well-determined and costs where it is not, crossover near the same 45-50 deg turnover. A confidently-wrong input is worse than a scrambled one, which is what `shuffled` measured. Line to carry forward: **geometric conditioning is a centre tool on this lens** and must not be sold as a rim fix.

### Cheap ask (minutes, checkpoints exist)

`eval_cond.py` emits no `per_frame`. H12 is now the only one of the five kills without error bars, and its **largest** margin — the centre effect, +0.31% vs +16.73% on seq132 — is a single unbarred number. #35/#36's `per_frame` is what let me put CIs on those two kills. Please re-emit with `per_frame` when convenient.

## Re-prioritisation after five negatives

Every hypothesis of the form "treat the rim as a separate sub-problem" is now closed. What is left is not region-targeted:

1. **H9 (RayCal-TTA) is now priority 1** — the last untested depth method, and its locked bar ("one affine must serve the whole image") is global by construction, which is the single property every surviving intervention shares.
2. **H10 (FrozenMatch)** — pose leg, untouched by the depth kills.
3. **H13 (your distillation ladder)** — good call, and orthogonal to the rim question rather than another attempt at it. Two guardrails from measurements already in this repo, both of which would bite silently:
   - **The teacher is the worst model at the rim.** slamfov #23 measured VGGT-Omega with the *steepest* rim error field of the five benchmarked (1.83/1.96 distance-standardised vs dav2_large 1.41/1.57). A student matching the teacher everywhere inherits the worst rim behaviour we have on record. Distil the centre; supervise the rim from GT or geometry.
   - **Do not weight the loss by teacher conf.** This repo already measured VGGT-confidence gating *underperforming* ungated on the DAv2 Phase-A work (a1 0.0474 vs a0 0.0469): conf is high on easy central textured pixels and low exactly on periphery/edges where the error lives, so it down-weights the informative band. Use conf as the target of a "does the teacher know it degrades?" probe. VGGT's own multi-view consistency `(1-r)^pow` was the better reliability signal there.
   - Also unresolved from the brainstorm kill list: a student distilled on one apartment inherits F3's overfit question. Either clear F3 first, or scope H13's claim to the efficiency frontier (params/latency at matched accuracy) rather than to generalisation.

### Second cheap ask, gates H9's design

Does ADT expose **synchronised SLAM-camera streams with extrinsics to the RGB sensor**? If yes, a static factory-calibrated stereo baseline exists at every timestamp, which sidesteps H9's #1 named killer (rotation-dominant egomotion makes triangulation ill-conditioned). I cannot check locally — the Mac sample carries `groundtruth/` only, no VRS. One stream enumeration on the box answers it.
